printing/printer.ml

```diff
@@ -635,34 +635,32 @@ let () =
       optwrite = (fun v -> should_print_dependent_evars := v) }
 
 let print_dependent_evars gl sigma seeds =
-  let rec pr_map sigma = function
-  | [] -> mt ()
-  | (evk, evi)::rest ->
-    let sep = if rest <> [] then ", " else "" in
-    (Evar.print evk) ++ str " : " ++ (Termops.pr_existential_key sigma evk) ++ str sep ++
-         (pr_map sigma rest)
-  in
   if !should_print_dependent_evars then
     let evars = Evarutil.gather_dependent_evars sigma seeds in
     let evars_pp = Evar.Map.fold (fun e i s ->
         let e' = pr_internal_existential_key e in
-        let sep = if s = (str "") then "" else ", " in
-        s ++ str sep ++ e' ++
+        s ++ e' ++
             (match i with
-            | None -> str " open"
+            | None -> (str " : ") ++ (Termops.pr_existential_key sigma e) ++ (str " open")
             | Some i ->
               let using = Evar.Set.fold (fun d s ->
-                  let sep = if s = (str "") then "" else " " in
-                  s ++ str sep ++ (pr_internal_existential_key d))
+                  s ++ str " " ++ (pr_internal_existential_key d))
                 i (str "") in
-              str " using " ++ using))
+              str " using" ++ using)
+          ++ str ", ")
       evars (str "")
     in
-    let map_evars = List.filter (fun b -> let (ev, _) = b in Evar.Map.mem ev evars)
-        (Evar.Map.bindings (Evd.undefined_map sigma)) in
+    let evars_current_pp = match gl with
+        | None -> str "no current goal"
+        | Some gl ->
+           let evars_current = Evarutil.gather_dependent_evars sigma [ gl ] in
+           Evar.Map.fold (fun e _ s ->
+               s ++ (pr_internal_existential_key e) ++ (str " "))
+             evars_current (str "current goal: ")
```

Comment:
1 The output after the `eapply strange_imp_trans` step is:

```
(dependent evars: ?X4 : ?P open, ?X5 : ?Q open, ; current goal: ?X5 )
```

However, the current goal is actually `?Goal`.  `?X5 = ?Q` is not.  Is the label `current goal` wrong or is the value wrong?

```
eapply strange_imp_trans.
4 focused subgoals
(shelved: 2) (ID 6)

  P1, P2, P3, P4 : Prop
  p12 : P1 -> P2
  p123 : (P1 -> P2) -> P3
  p34 : P3 -> P4
  ============================
  ?Q -> P4

subgoal 2 (ID 7) is:
 ?P -> ?Q
subgoal 3 (ID 8) is:
 ?P -> ?Q
subgoal 4 (ID 9) is:
 ?P

p14 < Show Existentials.
  <excerpted>
Existential 3 =
?Goal : [P1 : Prop
         P2 : Prop
         P3 : Prop
         P4 : Prop
         p12 : P1 -> P2
         p123 : (P1 -> P2) -> P3
         p34 : P3 -> P4 |- ?Q -> P4]
```

2 It's odd to see `fold` here, which suggests there could be multiple elements in the mapping.  There can only be one current goal.  Would be clearer with something that makes this clear, such as `Evar.Map.find_first`.  You might add a temporary check that `Evar.Map.cardinal` is 1.

Reply:
> However, the current goal is actually `?Goal`. `?X5 = ?Q` is not. Is the label `current goal` wrong or is the value wrong?

The evars after `current goal` are the evars used inside the current goal, see the feature description in #10420. This is not the current goal.

> 2 It's odd to see `fold` here, which suggests there could be multiple elements in the mapping. There can only be one current goal.

Please read my comment in PR #10489 from Jul 24, 2019, 10:42 pm GMT+2, there you see an example with multiple evars after `current goal`. The code uses `Evarutil.gather_dependent_evars`, which does not return the current goal, but rather all the evars used in the evar argument list, which is something completely different in most cases.